### PR TITLE
Add `onWillClose` callback to Modal

### PR DESCRIPTION
**Pull request by ticky**
Allows dynamically preventing the modal from closing.
Also, always passes real events through to the close handler.

### CODE REVIEW

File: packages/modal/example.jsx [hunk 1]
```diff
@@ -20,6 +20,16 @@ module.exports = React.createClass({
     });
   },
 
+  handleWillClose(evt) {
+    if (evt instanceof KeyboardEvent) {
```
Comment (nelix) [inline packages/modal/example.jsx]: is this like, ignoring esc?

Comment (ticky) [inline packages/modal/example.jsx]: It’s accepting esc implicitly. Just as an example of what you can do with it.

Comment (ticky) [inline packages/modal/example.jsx]: i.e. esc always closes, close button prompts.

**Comment by nelix** [comment]
Looks pretty legit. :+1:

**Comment by ticky** [comment]
@zendy: any thoughts on this? If you think it’s good, go ahead and merge it!

**Comment by zendy** [comment]
@ticky actually without context I can't say if I understand correctly. But I think the code looks good. So I'll merge it now.

**Pull request merged**